config/codec.go

```diff
@@ -0,0 +1,102 @@
+package config
+
+import (
+	"context"
+	"time"
+
+	"github.com/icza/dyno"
+	"github.com/tsaikd/KDGoLib/errutil"
+	"github.com/tsaikd/gogstash/config/logevent"
+)
+
+// errors
+var (
+	ErrorUnknownCodecType1 = errutil.NewFactory("unknown codec config type: %q")
+	ErrorInitCodecFailed1  = errutil.NewFactory("initialize codec module failed: %v")
+)
+
+// TypeCodecConfig is interface of codec module
+type TypeCodecConfig interface {
+	TypeCommonConfig
+	// The codec’s decode method is where data coming in from an input is transformed into an event.
+	Decode(ctx context.Context, data []byte, extra map[string]interface{}, msgChan chan<- logevent.LogEvent) (ok bool, err error)
+	// The encode method takes an event and serializes it (encodes) into another format.
+	Encode(ctx context.Context, event logevent.LogEvent, dataChan chan<- []byte) (ok bool, err error)
+}
+
+// CodecConfig is basic codec config struct
+type CodecConfig struct {
+	CommonConfig
+}
+
+// CodecHandler is a handler to regist codec module
+type CodecHandler func(ctx context.Context, raw *ConfigRaw) (TypeCodecConfig, error)
+
+var (
+	mapCodecHandler = map[string]CodecHandler{}
+)
+
+// RegistCodecHandler regist a codec handler
+func RegistCodecHandler(name string, handler CodecHandler) {
+	mapCodecHandler[name] = handler
+}
+
+// GetCodec returns a codec based on the 'codec' configuration from provided 'ConfigRaw' input
+func GetCodec(ctx context.Context, raw ConfigRaw) (TypeCodecConfig, error) {
+	codecConfig, _ := dyno.Get(map[string]interface{}(raw), "codec")
```

Comment:
Why not catch error?
> codecConfig, `err` := dyno.Get(map[string]interface{}(raw), "codec")

Reply:
There is no way that the error will occur, as we know for a fact that we are passing in a `map[string]interface{}`, so internally all it will do is to remap the structure to a nested `map[string]interface{}` instead of `map[interface]interface{}`  But I would be happy to add the check if needed.

Reply:
So this line definitely does not need error checking, because the `codec` config is optional, and if it is not present, the `dyno.Get` call will return an error, and we don't want to throw it... I can log it as an `Info`, but that's about it.  Want me to do that?

Reply:
The default behavior is delegated to the specific `input` implementation... Not everything can/will have a default codec, so I didn't want to ALWAYS return a codec, but rather return `nil` if provided `ConfigRaw` does not have a `codec` key.